### DISCUSSION TITLE
Bridge Framework Between Space ROS and FPrime

### DISCUSSION THREAD

**Discussion by iftahnaf** [space-ros/space-ros#81] [Ideas]
Hi all,  
  
We at [Lulav.Space](https://lulav.space/) would like to share with you our new [bridge framework](https://github.com/lulav/ros2_fprime_bridge) that connects ROS 2 and [FPrime](https://github.com/nasa/fprime).
We think this capability will be useful for Space ROS and we would love to get your take on this.
  
## Why Bridge ROS 2 and FPrime?  
  
- **ROS 2**: Offers a vast ecosystem and a set of functionalities that have become the standard in robotic applications.  
- **FPrime**: Provides a component-driven framework designed for real-time, embedded systems like satellites and spacecraft.  
  
## How it works

The user needs to define the messages going back and forth between the two frameworks by writing cpp headers that define the structures that make up the messages.  
  
The cpp headers are used as the source by which all communication messages are defined for all interface description languages. A Python script parses the headers, and generates schema files for:

  
- FPrime (`.fpp`)  
- ROS (`.msg`)  
- Protobuf (`.proto`, version 2 and 3)  
  
The user needs to define an FPrime component and a ROS node that will act as the two ends of the bridge. Communication between the FPrime component and the ROS node is performed via either UDP or named pipes (FIFO), for which simple wrapper classes are provided. Alternatively, you may choose any other relevant protocol.  
  
ProtoBuf serialization + deserialization takes (much) less than 100 microseconds (even for large messages with dozens of fields). On each end of the bridge, message fields are “translated” seriatim between Protobuf and ROS/FPrime. The Python script also provides automatic code generation for the translator functions (currently FPrime-ProtoBuf only).  
  
## Use Case  
  
We offer a simple example featuring a spring-mass-damper ROS 2 node, controlled by an FPrime controller component. This example provides a working model of how the bridge can facilitate a complex simulation that includes ROS 2 and FPrime components.  
  
## Future Plans  
  
- Replace ROS 2 with Space ROS.
- Integrate [ogma](https://github.com/nasa/ogma) capabilities to monitor specific signals from the dynamics (ROS 2) and/or controller (FPrime).  
- Integrate the simulation framework with CITROS for CI/CD and Monte Carlo simulations.

**Comment by Bckempa** [comment]
This is exciting and clearly relevant, thanks for sharing @iftahnaf.

@EzraBrooks, we should put this on the meeting agenda to discuss as a potential demo case.

**Reply by EzraBrooks** [reply]
Absolutely, this is awesome. Tools like this and BRASH (TRAClabs's cFS-to-ROS adapter) will be invaluable in Space ROS adoption, imo.

**Comment by ivanperez-keera** [comment]
Follow up regarding my comment from the last meeting:

We had a discussion at NASA about this and Luvav's situation and NPR7150.

With some additional checks (2 approvers for PRs and static analysis especially for cyber security checks), this could be distributed as part of Space ROS if we wanted.

If any other constraints emerge in the future, I'll let you know.

We can discuss further next week.

**Reply by ivanperez-keera** [reply]
Tagging @iftahnaf  for awareness.

**Reply by iftahnaf** [reply]
@ivanperez-keera Thanks for the update, I am looking forward to discussing it on Thursday.
